src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientCache.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.mongodb.spark.sql.connector.connection;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.jetbrains.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.client.MongoClient;
+
+import com.mongodb.spark.sql.connector.assertions.Assertions;
+
+/**
+ * A simple cache that handles the acquisition and release of {@code MongoClients}.
+ *
+ * <p>Handles the creation and closing of {@code MongoClients}. When a {@link MongoClient} is no
+ * longer used there is a grace period allowing the {@code MongoClient} to be reused. As Spark
+ * Executors can operate on multiple tasks this reduces the cost creating a MongoClient and
+ * connections.
+ */
+final class MongoClientCache {
```

Comment:
This code is much more simple than the previous connectors cache. It used as a Singleton and locks on acquisition and release of clients. This shouldn't be impediment as both are cheap operations.